h3/src/ext.rs

```diff
@@ -0,0 +1,104 @@
+//! Extensions for the HTTP/3 protocol.
+
+use std::convert::TryFrom;
+use std::str::FromStr;
+
+use bytes::{Buf, Bytes};
+
+use crate::{
+    error::Code,
+    proto::{stream::StreamId, varint::VarInt},
+    Error,
+};
+
+/// Describes the `:protocol` pseudo-header for extended connect
+///
+/// See: [https://www.rfc-editor.org/rfc/rfc8441#section-4]
+#[derive(Copy, PartialEq, Debug, Clone)]
+pub struct Protocol(ProtocolInner);
+
+impl Protocol {
+    /// WebTransport protocol
+    pub const WEB_TRANSPORT: Protocol = Protocol(ProtocolInner::WebTransport);
+}
+
+#[derive(Copy, PartialEq, Debug, Clone)]
+enum ProtocolInner {
+    WebTransport,
+}
+
+/// Error when parsing the protocol
+pub struct InvalidProtocol;
+
+impl FromStr for Protocol {
+    type Err = InvalidProtocol;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "webtransport" => Ok(Self(ProtocolInner::WebTransport)),
+            _ => Err(InvalidProtocol),
+        }
+    }
+}
+
+/// HTTP datagram frames
+/// See: https://www.rfc-editor.org/rfc/rfc9297#section-2.1
+pub struct Datagram<B = Bytes> {
+    /// Stream id divided by 4
+    stream_id: StreamId,
+    /// The data contained in the datagram
+    pub payload: B,
```

Comment:
Should this field be private?

Reply:
It is to allow destructuring, but an `into_payload(self) -> B` and `payload(&self) -> &B` should suffice, but it doesn't allow partial destructuring